translator/trace/otlp_to_oc.go

```diff
@@ -0,0 +1,400 @@
+	}
+	return ocAttributes
+}
+
+func attributeValueToString(attr *otlpcommon.AttributeKeyValue) string {
+	switch attr.Type {
+	case otlpcommon.AttributeKeyValue_STRING:
+		return attr.StringValue
+	case otlpcommon.AttributeKeyValue_BOOL:
+		return strconv.FormatBool(attr.BoolValue)
+	case otlpcommon.AttributeKeyValue_DOUBLE:
+		return strconv.FormatFloat(attr.DoubleValue, 'f', -1, 64)
+	case otlpcommon.AttributeKeyValue_INT:
+		return strconv.FormatInt(attr.IntValue, 10)
+	default:
+		return fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type)
+	}
+}
+
+func attributeValueToOC(attr *otlpcommon.AttributeKeyValue) *octrace.AttributeValue {
+	a := &octrace.AttributeValue{}
+
+	switch attr.Type {
+	case otlpcommon.AttributeKeyValue_STRING:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: attr.StringValue,
+			},
+		}
+	case otlpcommon.AttributeKeyValue_BOOL:
+		a.Value = &octrace.AttributeValue_BoolValue{
+			BoolValue: attr.BoolValue,
+		}
+	case otlpcommon.AttributeKeyValue_DOUBLE:
+		a.Value = &octrace.AttributeValue_DoubleValue{
+			DoubleValue: attr.DoubleValue,
+		}
+	case otlpcommon.AttributeKeyValue_INT:
+		a.Value = &octrace.AttributeValue_IntValue{
+			IntValue: attr.IntValue,
+		}
+	default:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type),
+			},
+		}
+	}
+
+	return a
+}
+
+// OTLP follows the W3C format, e.g. "vendorname1=opaqueValue1,vendorname2=opaqueValue2"
+// See https://w3c.github.io/trace-context/#combined-header-value
+func tracestateToOC(tracestate string) *octrace.Span_Tracestate {
+	if tracestate == "" {
+		return nil
+	}
+
+	// key-value pairs in the "key1=value1" format
+	pairs := strings.Split(tracestate, ",")
+
+	entries := make([]*octrace.Span_Tracestate_Entry, 0, len(pairs))
+	for _, pair := range pairs {
+		kv := strings.SplitN(pair, "=", 2)
+		key := kv[0]
+		val := ""
+		if len(kv) >= 2 {
+			val = kv[1]
+		}
+
+		entries = append(entries, &octrace.Span_Tracestate_Entry{
+			Key:   key,
+			Value: val,
+		})
+	}
+
+	return &octrace.Span_Tracestate{
+		Entries: entries,
+	}
+}
+
+func kindToOC(kind otlptrace.Span_SpanKind) octrace.Span_SpanKind {
+	switch kind {
+	case otlptrace.Span_SERVER:
+		return octrace.Span_SERVER
+	case otlptrace.Span_CLIENT:
+		return octrace.Span_CLIENT
+	// NOTE: see `kindToAttribute` function for custom kinds
+	case otlptrace.Span_SPAN_KIND_UNSPECIFIED:
+	case otlptrace.Span_INTERNAL:
+	case otlptrace.Span_PRODUCER:
+	case otlptrace.Span_CONSUMER:
+	default:
+	}
+
+	return octrace.Span_SPAN_KIND_UNSPECIFIED
+}
+
+func kindToAttribute(kind otlptrace.Span_SpanKind) *octrace.AttributeValue {
+	var ocKind OpenTracingSpanKind
+	switch kind {
+	case otlptrace.Span_SERVER:
+		ocKind = OpenTracingSpanKindServer
+	case otlptrace.Span_CLIENT:
+		ocKind = OpenTracingSpanKindClient
```

Comment:
We should not set attribute for these cases since they are directly representable in OC SpanKind. It sets unnecessary attribute and pollutes the attributes map which is visible in processors. This is undesirable.